In [24]:
import pyterrier as pt
import pickle
import pandas as pd
import numpy as np

In [25]:
if not pt.java.started():
    pt.java.init()

In [59]:
df = pd.read_json("./hiphop.json")
df

title  \
0                        20 Jahre Hiphop in Deutschland   
1                           The hiphop virtual machine.   
2                          The HipHop compiler for PHP.   
3     Fear of a Kanak Planet – Hiphop zwischen Weltk...   
4                          Q&A: From Esterel to HipHop.   
...                                                 ...   
1929  The energy cost of walking with arthritis of t...   
1930  A general method for determining the masses of...   
1931       ASCR/HEP Exascale Requirements Review Report   
1932  Superpartner mass measurements with 1D decompo...   
1933  Extensibility of hip adductors in children wit...   

                                                authors  year  \
0                           [Sascha Verlan, Hannes Loh]  2000   
1     [Keith Adams, Jason Evans, Bertrand Maher, Gui...  2014   
2     [Haiping Zhao, Iain Proctor, Minghui Yang, Xin...  2012   
3                            [Hannes Loh, Murat Güngör]  2002   
4                                        [Gregory Mone]  2014   
...                                                 ...   ...   
1929  [R. L. Waters, J. Perry, P. Conaty, B. Lunsfor...  1987   
1930            [Konstantin T. Matchev, Myeonghun Park]  2009   
1931  [Salman Habib, Robert Roser, Richard Gerber, K...  2016   
1932  [Partha Konar, Kyoungchul Kong, Konstantin T. ...  2009   
1933  [A. Lespargot, E. Renaudin, N. Khouri, M. Robert]  1994   

                                   editors               isbn publisher  \
0                                     None               None  Hannibal   
1     [Andrew P. Black, Todd D. Millstein]  978-1-4503-2585-1       ACM   
2      [Gary T. Leavens, Matthew B. Dwyer]  978-1-4503-1561-6       ACM   
3                                     None               None  Hannibal   
4                                     None               None      None   
...                                    ...                ...       ...   
1929                                  None               None      None   
1930                                  None               None      None   
1931                                  None               None      None   
1932                                  None               None      None   
1933                                  None               None      None   

                                               abstract  \
0                                                  None   
1                                                  None   
2                                                  None   
3                                                  None   
4                                                  None   
...                                                 ...   
1929  The physiologic energy expenditure of walking ...   
1930                                               None   
1931  This draft report summarizes and details the f...   
1932                                               None   
1933  The passive extension of relaxed hip adductor ...   

                                                   tags  \
0     [Geschichte-Gender, Hit_der_Geschichtswissensc...   
1                                                 [php]   
2                                                 [php]   
3     [Hit_der_Geschichtswissenschaft, {Jugendkultur...   
4                                                [dblp]   
...                                                 ...   
1929                      [energy, crutches, arthritis]   
1930                               [susy, physics, mt2]   
1931                                     [hpc,, review]   
1932                               [susy, physics, mt2]   
1933  [Child;, Humans;, Joint;, Palsy;, Female;, Cer...   

                                                    url  \
0                                                         
1     http://dblp.uni-trier.de/db/conf/oopsla/oopsla...   
2     http://dblp.uni-trier.de/db/conf/oopsla/oopsla...   
3    

In [60]:
df['docno'] = df['id']
df['text'] = df['title']

### Index mit mehreren Feldern anlegen

In [61]:
df_dict = df.to_dict(orient='records')

In [63]:
index_folder = "./index_hiphop"

fields=['text', 'tags', 'authors', 'abstract']

indexer = pt.IterDictIndexer(index_folder, meta={'docno': 420, 'text': 4200}, overwrite=True)
index_ref = indexer.index(df_dict, fields=fields)

print(f"Index-Pfad: {index_ref.toString()}")

Index-Pfad: ./index_hiphop/data.properties


In [64]:
pickle.dump(df, open("./hiphop_search.pkl", "wb"))

In [65]:
index = pt.IndexFactory.of(index_ref)

## Statistiken, etc.

In [66]:
print(index.getCollectionStatistics().toString())

Number of documents: 1934
Number of terms: 11594
Number of postings: 72031
Number of fields: 4
Number of tokens: 101043
Field names: [text, tags, authors, abstract]
Positions:   false



In [67]:
df.columns

Index(['title', 'authors', 'year', 'editors', 'isbn', 'publisher', 'abstract',
       'tags', 'url', 'id', 'type', 'search term', 'docno', 'text'],
      dtype='object')

In [68]:
for kv in index.getLexicon():
    print(f"{kv.getKey()} -> {kv.getValue().toString()}")

0 -> term1687 Nt=112 TF=273 maxTF=11 @{0 0 0} TFf=6,1,0,266
000 -> term3173 Nt=2 TF=2 maxTF=1 @{0 207 4} TFf=0,0,0,2
0001 -> term7528 Nt=2 TF=3 maxTF=2 @{0 213 4} TFf=0,0,0,3
001 -> term3668 Nt=12 TF=15 maxTF=3 @{0 219 2} TFf=0,0,0,15
0017 -> term1666 Nt=1 TF=1 maxTF=1 @{0 245 6} TFf=0,0,0,1
002 -> term8430 Nt=2 TF=3 maxTF=2 @{0 248 1} TFf=0,0,0,3
003 -> term10841 Nt=1 TF=1 maxTF=1 @{0 253 7} TFf=0,0,0,1
005 -> term8871 Nt=1 TF=1 maxTF=1 @{0 257 2} TFf=0,0,0,1
006 -> term9817 Nt=2 TF=2 maxTF=1 @{0 260 5} TFf=0,0,0,2
008 -> term7778 Nt=1 TF=1 maxTF=1 @{0 266 7} TFf=0,0,0,1
009 -> term8784 Nt=2 TF=2 maxTF=1 @{0 270 2} TFf=0,0,0,2
01 -> term7056 Nt=10 TF=13 maxTF=3 @{0 276 6} TFf=0,0,0,13
013 -> term11455 Nt=1 TF=2 maxTF=2 @{0 299 0} TFf=0,0,0,2
018 -> term11507 Nt=1 TF=1 maxTF=1 @{0 302 5} TFf=0,0,0,1
01a35 -> term7188 Nt=1 TF=1 maxTF=1 @{0 306 0} TFf=0,1,0,0
02 -> term7532 Nt=2 TF=2 maxTF=1 @{0 309 3} TFf=0,0,0,2
021 -> term11497 Nt=1 TF=1 maxTF=1 @{0 315 1} TFf=0,0,0,1
022 -> term8762 

In [69]:
term = "polit"
index.getLexicon()[term].toString()

'term885 Nt=11 TF=14 maxTF=2 @{0 103506 2} TFf=4,1,0,9'

In [70]:
term_freq_dict = {}

for kv in index.getLexicon():
    term_freq_dict[kv.getKey()] = kv.getValue().frequency

In [71]:
term_freq_dict_sorted = sorted(term_freq_dict.items(), key=lambda x: x[1], reverse=True)

In [72]:
term_freq_dict_sorted[:42]

[('hip', 1492),
 ('null', 1015),
 ('theori', 823),
 ('bishop', 687),
 ('dblp', 670),
 ('lattic', 669),
 ('2', 554),
 ('cerebr', 549),
 ('palsi', 540),
 ('model', 536),
 ('studi', 473),
 ('patient', 456),
 ('result', 455),
 ('1', 411),
 ('disloc', 381),
 ('black', 350),
 ('3', 333),
 ('can', 330),
 ('effect', 328),
 ('gener', 324),
 ('field', 318),
 ('us', 308),
 ('function', 307),
 ('gaug', 303),
 ('oper', 294),
 ('group', 285),
 ('4', 275),
 ('0', 273),
 ('method', 273),
 ('mass', 272),
 ('siphon', 267),
 ('base', 255),
 ('control', 255),
 ('child', 253),
 ('fermion', 251),
 ('hole', 249),
 ('children', 247),
 ('string', 247),
 ('symmetri', 247),
 ('qcd', 245),
 ('system', 243),
 ('show', 239)]

## TF-IDF-Suche

In [73]:
search_engine_tfidf = pt.terrier.Retriever(index, wmodel="TF_IDF", num_results=2000)

In [87]:
search_engine_tfidf.search("dance music")

qid  docid                                              docno  rank  \
0    1    159  https://www.bibsonomy.org/bibtex/2e8b5c548a89c...     0   
1    1    987  https://www.bibsonomy.org/bibtex/270a54c6051ab...     1   
2    1    944  https://www.bibsonomy.org/bibtex/2ffd8b801588c...     2   
3    1    977  https://www.bibsonomy.org/bibtex/241438a7150ff...     3   
4    1   1296  https://www.bibsonomy.org/bibtex/20a9593f3d71b...     4   
5    1     76  https://www.bibsonomy.org/bibtex/2abae0a38f92b...     5   
6    1    967  https://www.bibsonomy.org/bibtex/25ffa48a54335...     6   
7    1   1297  https://www.bibsonomy.org/bibtex/2146e68c8b0e0...     7   
8    1    947  https://www.bibsonomy.org/bibtex/2c13ba099506a...     8   
9    1    945  https://www.bibsonomy.org/bibtex/2bc006a682772...     9   
10   1    946  https://www.bibsonomy.org/bibtex/2f34832f94c0d...    10   
11   1    960  https://www.bibsonomy.org/bibtex/2beab2fb29b86...    11   
12   1    453  https://www.bibsonomy.org/bibtex/2bc4935df8feb...    12   
13   1    964  https://www.bibsonomy.org/bibtex/229bea0639fa0...    13   
14   1   1326  https://www.bibsonomy.org/bibtex/2f94ef053ae63...    14   

       score        query  
0   7.855363  dance music  
1   6.688628  dance music  
2   6.549164  dance music  
3   6.409024  dance music  
4   6.394595  dance music  
5   6.312176  dance music  
6   6.274756  dance music  
7   6.196511  dance music  
8   6.083580  dance music  
9   5.813347  dance music  
10  5.813347  dance music  
11  5.813347  dance music  
12  5.749087  dance music  
13  5.624736  dance music  
14  4.782129  dance music

### P und R

P = 0,57

R = 8/REL


In [19]:
8/14

0.5714285714285714

## TF-Suche

In [85]:
search_engine_tf = pt.BatchRetrieve(index, wmodel="Tf")

/tmp/ipykernel_3793/3266775375.py:1: DeprecationWarning: Call to deprecated class BatchRetrieve. (use pt.terrier.Retriever() instead) -- Deprecated since version 0.11.0.
  search_engine_tf = pt.BatchRetrieve(index, wmodel="Tf")


In [86]:
search_engine_tf.search("dance music")

qid  docid                                              docno  rank  score  \
0    1     76  https://www.bibsonomy.org/bibtex/2abae0a38f92b...     0    7.0   
1    1   1296  https://www.bibsonomy.org/bibtex/20a9593f3d71b...     1    5.0   
2    1   1297  https://www.bibsonomy.org/bibtex/2146e68c8b0e0...     2    5.0   
3    1    159  https://www.bibsonomy.org/bibtex/2e8b5c548a89c...     3    2.0   
4    1    987  https://www.bibsonomy.org/bibtex/270a54c6051ab...     4    2.0   
5    1   1326  https://www.bibsonomy.org/bibtex/2f94ef053ae63...     5    2.0   
6    1    453  https://www.bibsonomy.org/bibtex/2bc4935df8feb...     6    1.0   
7    1    944  https://www.bibsonomy.org/bibtex/2ffd8b801588c...     7    1.0   
8    1    945  https://www.bibsonomy.org/bibtex/2bc006a682772...     8    1.0   
9    1    946  https://www.bibsonomy.org/bibtex/2f34832f94c0d...     9    1.0   
10   1    947  https://www.bibsonomy.org/bibtex/2c13ba099506a...    10    1.0   
11   1    960  https://www.bibsonomy.org/bibtex/2beab2fb29b86...    11    1.0   
12   1    964  https://www.bibsonomy.org/bibtex/229bea0639fa0...    12    1.0   
13   1    967  https://www.bibsonomy.org/bibtex/25ffa48a54335...    13    1.0   
14   1    977  https://www.bibsonomy.org/bibtex/241438a7150ff...    14    1.0   

          query  
0   dance music  
1   dance music  
2   dance music  
3   dance music  
4   dance music  
5   dance music  
6   dance music  
7   dance music  
8   dance music  
9   dance music  
10  dance music  
11  dance music  
12  dance music  
13  dance music  
14  dance music

In [22]:
search_engine_tf.search("politics rap").docno[12]

'https://www.bibsonomy.org/bibtex/2684f23c73b5f1005d2a28744e3f1d8a9/whiteness'

### Berechnung IDF zum Prüfen der TF-IDF-Berechnung

In [28]:
from nltk.stem import PorterStemmer
stemmer = PorterStemmer()

In [29]:
def get_idf_for_term(term, index):
    lex = index.getLexicon()
    stemmed_term = stemmer.stem(term)

    if not stemmed_term in lex:
        return

    lex_entry = lex[stemmed_term]

    df_term = lex_entry.getDocumentFrequency()
    N = index.getCollectionStatistics().numberOfDocuments

    # inverse document frequency
    idf = N/df_term

    # apply logarithm(base 10) to idf
    log_idf = np.emath.logn(10, idf)

    return log_idf

In [30]:
term = "politics"
get_idf_for_term(term, index)

2.2450637845887575

## Code um TF-IDF selbst zu berechnen

#### funktioniert, im Vergleich zum selbst ausgerechneten TF-IDF, aber anderer Wert als search_engine_tfidf 

Beispiel: docid 82: 

score-TF: 4 -> rap=3, politics = 1

idf rap = 2.5083052193633386

idf politics = 2.2450637845887575

3 * 2.5083052193633386 + 1 * 2.2450637845887575 = 9.769979442678775

calc_tf_idf: 9.769979442678775

search_engine_tfidf: 10.612571

erster Tipp für den Unterschied wäre Rundungsdifferenzen, aber bei in der Mitte gerankten Ergebnissen ist die Differenz größer.

Bsp: docid 489
calc_: 2.2450637845887575
search-engine: 5.677485

#### Dieser Eintrag  ist als Ergebnis auch nicht relevant...


In [83]:
def calc_tf_idf(query, docno, index):

    lex = index.getLexicon()
    terms = query.split()
    tf_scores = []

    for term in terms:
        
        stemmed_term = stemmer.stem(term)
    
        if not stemmed_term in lex:
            
            continue
        
        lex_entry = lex[stemmed_term]
    
        df_term = lex_entry.getDocumentFrequency()
        N = index.getCollectionStatistics().numberOfDocuments
    
        idf = N/df_term    
        log_idf = np.emath.logn(10, idf)

        tf_term = search_engine_tf.search(term)[search_engine_tf.search(term)["docno"] == docno]["score"]

        if tf_term.empty:
            continue
            
        tf_int = int(tf_term)
        
        tf_idf = tf_int*log_idf

        tf_scores.append(tf_idf)
        
    return sum(tf_scores)

In [102]:
query = "politics rap"
docno = "https://www.bibsonomy.org/bibtex/2684f23c73b5f1005d2a28744e3f1d8a9/whiteness"

calc_tf_idf(query, docno, index)

/tmp/ipykernel_9800/3554385174.py:28: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead
  tf_int = int(tf_term)


2.2450637845887575